shared-brotli-patch-decoder/src/lib.rs

```diff
@@ -9,42 +9,77 @@ pub mod decode_error;
 #[cfg(fuzzing)]
 use decode_error::DecodeError;
 
-/// Decodes shared brotli encoded data using the optional shared dictionary.
-///
-/// The shared dictionary is a raw LZ77 style dictionary, see:
-/// <https://datatracker.ietf.org/doc/html/draft-vandevenne-shared-brotli-format-11#section-3.2>
-///
-/// Will fail if the decoded result will be greater then max_uncompressed_length. Any excess data
-/// in encoded after the encoded stream finishes is also considered an error.
-pub fn shared_brotli_decode(
-    encoded: &[u8],
-    shared_dictionary: Option<&[u8]>,
-    max_uncompressed_length: usize,
-) -> Result<Vec<u8>, decode_error::DecodeError> {
-    cfg_if::cfg_if! {
-        if #[cfg(fuzzing)]
-        {
-            // When running under a fuzzer disable brotli decoding and instead just pass through the input data.
-            // This allows the fuzzer to more effectively explore code gated behind brotli decoding.
-            // TODO(garretrieger): instead consider modifying the top level IFT apis to allow a custom brotli decoder
-            //   implementation to be provided. This would allow fuzzing to sub in a custom impl that could return all
-            //   of the possible errors that the standard impl here can generate.
-            return if encoded.len() <= max_uncompressed_length {
-                Ok(encoded.to_vec())
+/// A shared brotli decoder.
+pub trait SharedBrotliDecoder {
+    /// Decodes shared brotli encoded data using the optional shared dictionary.
+    ///
+    /// The shared dictionary is a raw LZ77 style dictionary, see:
+    /// <https://datatracker.ietf.org/doc/html/draft-vandevenne-shared-brotli-format-11#section-3.2>
+    ///
+    /// Will fail if the decoded result will be greater then max_uncompressed_length. Any excess data
+    /// in encoded after the encoded stream finishes is also considered an error.
+    fn decode(
+        &self,
+        encoded: &[u8],
+        shared_dictionary: Option<&[u8]>,
+        max_uncompressed_length: usize,
+    ) -> Result<Vec<u8>, decode_error::DecodeError>;
+}
+
+pub struct BuiltInBrotliDecoder;
+
+pub struct ExternalBrotliDecoder(pub Box<dyn SharedBrotliDecoder>);
+
+/// Just passes through the input data, primarily intended for use in fuzzers.
+pub struct NoopBrotliDecoder;
+
+impl SharedBrotliDecoder for ExternalBrotliDecoder {
+    fn decode(
+        &self,
+        encoded: &[u8],
+        shared_dictionary: Option<&[u8]>,
+        max_uncompressed_length: usize,
+    ) -> Result<Vec<u8>, decode_error::DecodeError> {
+        self.0
+            .decode(encoded, shared_dictionary, max_uncompressed_length)
+    }
+}
```

Comment:
this is a bit funny, I'm wondering why we are boxing the inner type instead of using it directly, given that it impls the trait?